adapters/repos/db/lsmkv/snapshot.go

```diff
@@ -0,0 +1,97 @@
+package lsmkv
+
+import (
+	"context"
+	"io/fs"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+)
+
+// PauseCompaction waits for all ongoing compactions to finish,
+// then makes sure that no new compaction can be started.
+//
+// This is a preparatory stage for taking snapshots.
+//
+// A timeout should be specified for the input context as some
+// compactions are long-running, in which case it may be better
+// to fail the backup attempt and retry later, than to block
+// indefinitely.
+func (b *Bucket) PauseCompaction(ctx context.Context) error {
+	compactionHalted := make(chan struct{})
+
+	go func() {
+		b.disk.stopCompactionCycle <- struct{}{}
+		compactionHalted <- struct{}{}
+	}()
+
+	select {
+	case <-ctx.Done():
+		// resume the compaction cycle, as the
+		// context deadline was exceeded
+		defer b.disk.initCompactionCycle(DefaultCompactionInterval)
```

Comment:
I'm afraid this may start 2nd compaction goroutine before first one finishes. If the original one is still running, this will start after timeout and then one of them will be stopped.
I wonder if `<- b.disk.stopCompactionCycle` here would do the trick - get stop from channel before compaction cycle would read it.

Reply:
resolved by the addition of CycleManager to flush and compaction cycles